### PR TITLE
[SW-2120][backport-3.28.1] Enable H2O to Connect to Hive

### PR DESCRIPTION
**Pull request by mn-mikke**

### CODE REVIEW

File: core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala [hunk 1]
```diff
@@ -145,8 +147,10 @@ trait ExternalBackendConf extends SharedBackendConf with Logging {
 
   def setKillOnUnhealthyClusterDisabled(): H2OConf = set(PROP_EXTERNAL_CLUSTER_KILL_ON_UNHEALTHY._1, value = false)
 
+  @DeprecatedMethod("""set("spark.yarn.principal", principal)""", "3.32")
```
Comment (mn-mikke) [inline core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala]: @jakubhava I will have to remove the deprecation and keep `kerberosPrincipal` and `kerberosKeytab`. Spark doesn't allow to specify a proxy user and a principal at the same time. In Spark with Livy, Livy authenticates the principal with keytab and  sets proxy user via `klist`.

Comment (jakubhava) [reply core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala]: oki oki, thanks for the info! I guess it applies also for the PR to master right?

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalH2OBackend.scala [hunk 2]
```diff
@@ -182,4 +188,20 @@ object InternalH2OBackend extends InternalBackendUtils {
...
+        conf.hiveJdbcUrlPattern.get,
+        conf.hiveHost.get,
+        conf.hivePrincipal.get)
+
+      val principal = conf.kerberosPrincipal.get
+      val keytabPath = conf.kerberosKeytab.get
+      val configuration = new Configuration()
+      configuration.set(DelegationTokenRefresher.H2O_HIVE_JDBC_URL, jdbcUrl)
+      configuration.set(DelegationTokenRefresher.H2O_HIVE_PRINCIPAL, conf.hivePrincipal.get)
+      val options = HiveTokenGenerator.HiveOptions.make(configuration)
```
Comment (honzasterba) [inline core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalH2OBackend.scala]: you can call make(url, pricipal)

File: core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala (hunk 1 above)
Comment (mn-mikke) [reply core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala]: yes, you are right.

File: doc/src/site/sphinx/migration_guide.rst [hunk 3]
```diff
@@ -29,6 +29,12 @@ From 3.30 to 3.32
   use ``H2OGBM`` instead. The classes wrapping these methods are removed as well. This removal affects only Scala API
   as other APIs don't have such functionality.
 
+- The configuration option ``spark.ext.h2o.external.kerberos.principal`` was replaced
```
Comment (jakubhava) [inline doc/src/site/sphinx/migration_guide.rst]: 👍

**Comment by jakubhava** [comment]
@mn-mikke seems like we can close this PR and keep the one opened for master then?

**Comment by mn-mikke** [comment]
yes, we can.

File: core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala [hunk 4]
```diff
@@ -471,4 +493,23 @@ object SharedBackendConf {
...
+
+  /** The full address of HiveServer2, for example hostname:10000 */
+  val PROP_HIVE_HOST: (String, None.type) = ("spark.ext.h2o.hive.host", None)
+
+  /** Hiveserver2 Kerberos principal, for example hive/hostname@DOMAIN.COM */
+  val PROP_HIVE_PRINCIPAL: (String, None.type) = ("spark.ext.h2o.hive.principal", None)
+
+  /** Can be used to further customize the way the driver constructs the Hive JDBC URL */
+  val PROP_HIVE_JDBC_URL_PATTERN: (String, String) =
+    ("spark.ext.h2o.hive.jdbc_url_pattern", "jdbc:hive2://{{host}}/;{{auth}}")
```
Comment (honzasterba) [inline core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala:511]: do we need to duplicate the default value here?

Comment (mn-mikke) [reply core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala:511]: Where is the second definition? In H2O-3 codebase? Is this property optional https://github.com/h2oai/sparkling-water/pull/2000/files/353b70dd3f6362c1310ac8733dfff13c0823309a#diff-e3983f89666b52c7f5935a5ea1a6c29cR196 if we call `DelegationTokenRefresher.setup`?

Comment (honzasterba) [reply core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala:511]: yes, this property is optional. BUT: either urlPattern OR hiveHost must be set

Comment (honzasterba) [reply core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala:511]: default is defined
https://github.com/h2oai/h2o-3/blob/5161c70bff9e717d6c2f1b310598aa7de6437dd3/h2o-hive/src/main/java/water/hive/HiveTokenGenerator.java#L32
and here
https://github.com/h2oai/h2o-3/blob/5161c70bff9e717d6c2f1b310598aa7de6437dd3/h2o-hive/src/main/java/water/hive/HiveTokenGenerator.java#L45

Comment (mn-mikke) [reply core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala]: @honzasterba I've made the changes and the default values is not defined in SW anymore. Could you double check that [these conditions](https://github.com/h2oai/sparkling-water/pull/2000/files#diff-8007bc0a47fbb7c1aa706403041431f4R114-R129) are correct?

Comment (honzasterba) [reply core/src/main/scala/ai/h2o/sparkling/backend/shared/SharedBackendConf.scala]: lgtm

File: core/src/main/scala/ai/h2o/sparkling/backend/shared/ArgumentBuilder.scala [hunk 5]
```diff
@@ -76,7 +76,7 @@ class ArgumentBuilder() {
 
   def addIf(arg: String, value: Option[String], condition: Boolean): ArgumentBuilder = {
     if (condition) {
-      add(arg, value.get)
+      add(arg, value)
```
Comment (mn-mikke) [inline core/src/main/scala/ai/h2o/sparkling/backend/shared/ArgumentBuilder.scala:79]: this was a bug

Comment (jakubhava) [reply core/src/main/scala/ai/h2o/sparkling/backend/shared/ArgumentBuilder.scala:79]: Good catch!, it's fixed on master

**Pull request merged**